File: services/api/app/services/graph_intelligence_service.py
"""
Graph Intelligence Service
Handles substitutions, confusion disambiguation, and ingredient pairings
"""

from typing import List, Dict, Any, Optional, Set
from datetime import datetime
import asyncpg
from collections import defaultdict


class GraphIntelligenceService:
    """Service for graph-based ingredient intelligence"""
    
    def __init__(self):
        """Initialize graph intelligence service"""
        pass
    
    async def get_substitutions(
        self,
        conn,
        ingredient_id: str,
        context: Optional[Dict[str, Any]] = None,
        limit: int = 10
    ) -> List[Dict[str, Any]]:
        """
        Find best substitutes for an ingredient
        
        Args:
            conn: Database connection
            ingredient_id: Source ingredient UUID
            context: Optional context (dish_type, cuisine, dietary_restrictions, form)
            limit: Maximum number of results
            
        Returns:
            List of substitution suggestions with scores and context
        """
        try:
            # Build query with optional context filters
            sql = """
                SELECT 
                    sub.id,
                    sub.target_ingredient_id,
                    mi.canonical_name,
                    mi.category,
                    mi.names,
                    mi.taste_profile,
                    mi.common_uses,
                    sub.substitution_type,
                    sub.similarity_score,
                    sub.applicable_forms,
                    sub.applicable_dishes,
                    sub.notes,
                    sub.user_acceptance_rate,
                    sub.times_suggested,
                    sub.times_accepted
                FROM ingredient_substitutions sub
                JOIN master_ingredients mi ON sub.target_ingredient_id = mi.id
                WHERE sub.source_ingredient_id = $1
            """
            
            params = [ingredient_id]
            param_idx = 2
            
            # Add context filters
            if context:
                # Filter by applicable form (fresh, dried, powdered)
                if context.get("form"):
                    sql += f" AND $2 = ANY(sub.applicable_forms)"
                    params.append(context["form"])
                    param_idx += 1
                
                # Filter by dish type
                if context.get("dish_type"):
                    sql += f" AND ${param_idx} = ANY(sub.applicable_dishes)"
                    params.append(context["dish_type"])
                    param_idx += 1
            
            # Order by similarity and acceptance rate
            sql += """
                ORDER BY 
                    sub.similarity_score DESC,
                    sub.user_acceptance_rate DESC NULLS LAST,
                    sub.times_accepted DESC
                LIMIT $""" + str(param_idx)
            params.append(limit)
            
            results = await conn.fetch(sql, *params)
            
            # Format results
            substitutions = []
            for row in results:
                substitutions.append({
                    "substitution_id": str(row["id"]),
                    "ingredient_id": str(row["target_ingredient_id"]),
                    "canonical_name": row["canonical_name"],
                    "category": row["category"],
                    "names": row["names"],
                    "taste_profile": row["taste_profile"],
                    "common_uses": row["common_uses"],
                    "substitution_type": row["substitution_type"],
                    "similarity_score": float(row["similarity_score"]),
                    "applicable_forms": row["applicable_forms"],
                    "applicable_dishes": row["applicable_dishes"],
                    "notes": row["notes"],
                    "user_acceptance_rate": float(row["user_acceptance_rate"]) if row["user_acceptance_rate"] else None,
                    "usage_stats": {
                        "times_suggested": row["times_suggested"],
                        "times_accepted": row["times_accepted"]
                    }
                })
            
            return substitutions
            
        except Exception as e:
            print(f"Error getting substitutions: {e}")
            return []
    
    async def resolve_confusion(
        self,
        conn,
        detected_ingredients: List[str],
        visual_features: Optional[Dict[str, Any]] = None,
        user_context: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Disambiguate between commonly confused ingredients
        
        Args:
            conn: Database connection
            detected_ingredients: List of ingredient IDs that might be confused
            visual_features: Optional visual features from image (colors, texture, etc.)
            user_context: Optional user context (location, cuisine preference)
            
        Returns:
            Disambiguation result with recommendations and differentiating factors
        """
        if len(detected_ingredients) < 2:
            return {
                "needs_disambiguation": False,
                "detected_count": len(detected_ingredients)
            }
        
        try:
            # Find confusion pairs
            confusion_data = []
            
            for i, ing_a in enumerate(detected_ingredients):
                for ing_b in detected_ingredients[i+1:]:
                    # Check if confusion exists
                    result = await conn.fetchrow("""
                        SELECT 
                            conf.id,
                            conf.confusion_reason,
                            conf.confusion_frequency,
                            conf.disambiguation_rules,
                            conf.key_visual_differences,
                            mi_a.canonical_name as ingredient_a_name,
                            mi_b.canonical_name as ingredient_b_name
                        FROM ingredient_confusion conf
                        JOIN master_ingredients mi_a ON conf.ingredient_a_id = mi_a.id
                        JOIN master_ingredients mi_b ON conf.ingredient_b_id = mi_b.id
                        WHERE (conf.ingredient_a_id = $1 AND conf.ingredient_b_id = $2)
                           OR (conf.ingredient_a_id = $2 AND conf.ingredient_b_id = $1)
                    """, ing_a, ing_b)
                    
                    if result:
                        confusion_data.append({
                            "confusion_id": str(result["id"]),
                            "ingredient_a": ing_a,
                            "ingredient_b": ing_b,
                            "ingredient_a_name": result["ingredient_a_name"],
                            "ingredient_b_name": result["ingredient_b_name"],
                            "confusion_reason": result["confusion_reason"],
                            "confusion_frequency": result["confusion_frequency"],
                            "disambiguation_rules": result["disambiguation_rules"],
                            "key_visual_differences": result["key_visual_differences"]
                        })
            
            if not confusion_data:
                return {
                    "needs_disambiguation": False,
                    "detected_count": len(detected_ingredients),
                    "message": "No known confusion patterns between detected ingredients"
                }
            
            # Apply disambiguation rules
            recommendations = []
            
            for confusion in confusion_data:
                recommendation = {
                    "confused_pair": [
                        confusion["ingredient_a_name"],
                        confusion["ingredient_b_name"]
                    ],
                    "reason": confusion["confusion_reason"],
                    "disambiguation_tips": confusion["disambiguation_rules"],
                    "visual_differences": confusion["key_visual_differences"],
                    "confidence": "medium"
                }
                
                # Apply visual feature matching if available
                if visual_features:
                    # Check dominant colors
                    if visual_features.get("dominant_colors"):
                        for diff in confusion["key_visual_differences"]:
                            if any(color in diff.lower() for color in visual_features["dominant_colors"]):
                                recommendation["matched_visual_clue"] = diff
                                recommendation["confidence"] = "high"
                
                recommendations.append(recommendation)
            
            return {
                "needs_disambiguation": True,
                "detected_count": len(detected_ingredients),
                "confusion_patterns": len(confusion_data),
                "recommendations": recommendations
            }
            
        except Exception as e:
            print(f"Error resolving confusion: {e}")
            return {
                "needs_disambiguation": False,
                "error": str(e)
            }
    
    async def get_pairings(
        self,
        conn,
        ingredient_ids: List[str],
        cuisine_type: Optional[str] = None,
        dish_type: Optional[str] = None,
        limit: int = 20
    ) -> List[Dict[str, Any]]:
        """
        Get complementary ingredient pairing suggestions
        
        Args:
            conn: Database connection
            ingredient_ids: List of ingredient IDs to find pairings for
            cuisine_type: Optional cuisine filter (indian, italian, chinese, etc.)
            dish_type: Optional dish type filter (curry, pasta, stir_fry, etc.)
            limit: Maximum number of suggestions
            
        Returns:
            List of ingredient pairing suggestions
        """
        if not ingredient_ids:
            return []
        
        try:
            # Get pairings for all provided ingredients
            sql = """
                SELECT 
                    pair.id,
                    pair.ingredient_a_id,
                    pair.ingredient_b_id,
                    mi_a.canonical_name as ingredient_a_name,
                    mi_b.canonical_name as ingredient_b_name,
                    mi_b.category as ingredient_b_category,
                    mi_b.names as ingredient_b_names,
                    mi_b.taste_profile,
                    pair.pairing_score,
                    pair.pairing_type,
                    pair.cuisine_types,
                    pair.dish_types,
                    pair.source,
                    pair.times_used_together
                FROM ingredient_pairings pair
                JOIN master_ingredients mi_a ON pair.ingredient_a_id = mi_a.id
                JOIN master_ingredients mi_b ON pair.ingredient_b_id = mi_b.id
                WHERE pair.ingredient_a_id = ANY($1)
                   OR pair.ingredient_b_id = ANY($1)
            """
            
            params = [ingredient_ids]
            param_idx = 2
            
            # Add cuisine filter
            if cuisine_type:
                sql += f" AND ${param_idx} = ANY(pair.cuisine_types)"
                params.append(cuisine_type)
                param_idx += 1
            
            # Add dish type filter
            if dish_type:
                sql += f" AND ${param_idx} = ANY(pair.dish_types)"
                params.append(dish_type)
                param_idx += 1
            
            sql += f"""
                ORDER BY 
                    pair.pairing_score DESC,
                    pair.times_used_together DESC
                LIMIT ${param_idx}
            """
            params.append(limit)
            
            results = await conn.fetch(sql, *params)
            
            # Deduplicate and format results
            seen_ingredients = set(ingredient_ids)
            pairings = []
            
            for row in results:
                # Determine which ingredient is the new suggestion
                if str(row["ingredient_a_id"]) in ingredient_ids:
                    suggested_id = str(row["ingredient_b_id"])
                    suggested_name = row["ingredient_b_name"]
                    suggested_category = row["ingredient_b_category"]
                    suggested_names = row["ingredient_b_names"]
                    suggested_taste = row["taste_profile"]
                else:
                    # Ingredient B is in our list, so A is the suggestion
                    continue  # Skip to avoid duplicates
                
                # Skip if already suggested
                if suggested_id in seen_ingredients:
                    continue
                
                seen_ingredients.add(suggested_id)
                
                pairings.append({
                    "pairing_id": str(row["id"]),
                    "ingredient_id": suggested_id,
                    "canonical_name": suggested_name,
                    "category": suggested_category,
                    "names": suggested_names,
                    "taste_profile": suggested_taste,
                    "pairing_score": float(row["pairing_score"]),
                    "pairing_type": row["pairing_type"],
                    "cuisine_types": row["cuisine_types"],
                    "dish_types": row["dish_types"],
                    "source": row["source"],
                    "times_used_together": row["times_used_together"]
                })
            
            return pairings
            
        except Exception as e:
            print(f"Error getting pairings: {e}")
            return []
    
    async def calculate_recipe_compatibility(
        self,
        conn,
        ingredient_ids: List[str],
        cuisine_type: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Calculate how well a set of ingredients work together
        
        Args:
            conn: Database connection
            ingredient_ids: List of ingredient IDs in recipe
            cuisine_type: Optional cuisine context
            
        Returns:
            Compatibility score and analysis
        """
        if len(ingredient_ids) < 2:
            return {
                "compatibility_score": 1.0,
                "ingredient_count": len(ingredient_ids),
                "message": "Need at least 2 ingredients for compatibility analysis"
            }
        
        try:
            # Get all pairings between ingredients
            pairing_scores = []
            pairing_details = []
            
            for i, ing_a in enumerate(ingredient_ids):
                for ing_b in ingredient_ids[i+1:]:
                    result = await conn.fetchrow("""
                        SELECT 
                            pair.pairing_score,
                            pair.pairing_type,
                            pair.cuisine_types,
                            mi_a.canonical_name as ingredient_a_name,
                            mi_b.canonical_name as ingredient_b_name
                        FROM ingredient_pairings pair
                        JOIN master_ingredients mi_a ON pair.ingredient_a_id = mi_a.id
                        JOIN master_ingredients mi_b ON pair.ingredient_b_id = mi_b.id
                        WHERE (pair.ingredient_a_id = $1 AND pair.ingredient_b_id = $2)
                           OR (pair.ingredient_a_id = $2 AND pair.ingredient_b_id = $1)
                    """, ing_a, ing_b)
                    
                    if result:
                        score = float(result["pairing_score"])
                        pairing_scores.append(score)
                        pairing_details.append({
                            "ingredients": [
                                result["ingredient_a_name"],
                                result["ingredient_b_name"]
                            ],
                            "score": score,
                            "type": result["pairing_type"],
                            "cuisines": result["cuisine_types"]
                        })
            
            # Calculate overall compatibility
            if not pairing_scores:
                compatibility_score = 0.5  # Neutral if no known pairings
                confidence = "low"
            else:
                # Average pairing score
                compatibility_score = sum(pairing_scores) / len(pairing_scores)
                confidence = "high" if len(pairing_scores) >= len(ingredient_ids) else "medium"
            
            # Determine compatibility level
            if compatibility_score >= 0.8:
                compatibility_level = "excellent"
            elif compatibility_score >= 0.6:
                compatibility_level = "good"
            elif compatibility_score >= 0.4:
                compatibility_level = "fair"
            else:
                compatibility_level = "poor"
            
            return {
                "compatibility_score": round(compatibility_score, 3),
                "compatibility_level": compatibility_level,
                "confidence": confidence,
                "ingredient_count": len(ingredient_ids),
                "known_pairings": len(pairing_scores),
                "pairing_details": pairing_details
            }
            
        except Exception as e:
            print(f"Error calculating recipe compatibility: {e}")
            return {
                "compatibility_score": 0.0,
                "error": str(e)
            }
    
    async def optimize_grocery_list(
        self,
        conn,
        ingredient_ids: List[str],
        user_inventory: Optional[List[str]] = None,
        budget_constraint: Optional[float] = None
    ) -> Dict[str, Any]:
        """
        Optimize grocery list by suggesting substitutions and consolidations
        
        Args:
            conn: Database connection
            ingredient_ids: List of ingredients needed
            user_inventory: Optional list of ingredient IDs already owned
            budget_constraint: Optional budget limit
            
        Returns:
            Optimized grocery list with suggestions
        """
        user_inventory = user_inventory or []
        
        try:
            # Identify ingredients already in inventory
            already_have = [ing for ing in ingredient_ids if ing in user_inventory]
            need_to_buy = [ing for ing in ingredient_ids if ing not in user_inventory]
            
            # Get ingredient details for items to buy
            if need_to_buy:
                results = await conn.fetch("""
                    SELECT 
                        id,
                        canonical_name,
                        category,
                        names,
                        common_uses
                    FROM master_ingredients
                    WHERE id = ANY($1)
                """, need_to_buy)
                
                items_to_buy = [dict(row) for row in results]
            else:
                items_to_buy = []
            
            # Find consolidation opportunities (multiple items of same category)
            category_groups = defaultdict(list)
            for item in items_to_buy:
                category_groups[item["category"]].append(item)
            
            consolidation_suggestions = []
            for category, items in category_groups.items():
                if len(items) > 1:
                    consolidation_suggestions.append({
                        "category": category,
                        "items": [item["canonical_name"] for item in items],
                        "suggestion": f"Consider buying these {len(items)} {category.lower()} items together"
                    })
            
            # Find substitution opportunities for expensive/rare items
            substitution_suggestions = []
            for item_id in need_to_buy:
                # Check if cheaper/more available substitutes exist
                substitutes = await self.get_substitutions(
                    conn,
                    item_id,
                    context={"substitution_type": "emergency"},
                    limit=3
                )
                
                if substitutes:
                    substitution_suggestions.append({
                        "ingredient_id": item_id,
                        "alternatives": [
                            {
                                "name": sub["canonical_name"],
                                "similarity": sub["similarity_score"],
                                "type": sub["substitution_type"]
                            }
                            for sub in substitutes[:2]  # Top 2 alternatives
                        ]
                    })
            
            return {
                "total_ingredients": len(ingredient_ids),
                "already_have": len(already_have),
                "need_to_buy": len(need_to_buy),
                "items_to_buy": items_to_buy,
                "optimizations": {
                    "consolidation_opportunities": consolidation_suggestions,
                    "substitution_suggestions": substitution_suggestions
                },
                "estimated_savings": len(substitution_suggestions)  # Placeholder
            }
            
        except Exception as e:
            print(f"Error optimizing grocery list: {e}")
            return {
                "error": str(e)
            }
    
    async def record_substitution_feedback(
        self,
        conn,
        substitution_id: str,
        was_accepted: bool,
        feedback_note: Optional[str] = None
    ) -> bool:
        """
        Record user feedback on substitution suggestion
        
        Args:
            conn: Database connection
            substitution_id: UUID of substitution
            was_accepted: Whether user accepted the substitution
            feedback_note: Optional feedback text
            
        Returns:
            Success boolean
        """
        try:
            if was_accepted:
                await conn.execute("""
                    UPDATE ingredient_substitutions
                    SET times_accepted = times_accepted + 1,
                        times_suggested = times_suggested + 1,
                        user_acceptance_rate = 
                            ROUND((times_accepted + 1.0) / (times_suggested + 1.0), 2)
                    WHERE id = $1
                """, substitution_id)
            else:
                await conn.execute("""
                    UPDATE ingredient_substitutions
                    SET times_suggested = times_suggested + 1,
                        user_acceptance_rate = 
                            ROUND(times_accepted / (times_suggested + 1.0), 2)
                    WHERE id = $1
                """, substitution_id)
            
            return True
            
        except Exception as e:
            print(f"Error recording feedback: {e}")
            return False

File: apps/mobile/lib/services/graph_intelligence_service.dart
"""
Graph Intelligence Service for Flutter Mobile App
Handles substitutions, confusions, pairings, and recipe compatibility
"""

import 'dart:convert';
import 'package:http/http.dart' as http;

class GraphIntelligenceService {
  final String baseUrl;
  
  GraphIntelligenceService({required this.baseUrl});
  
  /// Get substitution recommendations for an ingredient
  Future<SubstitutionResponse> getSubstitutions({
    required String ingredientId,
    SubstitutionContext? context,
    int limit = 10,
  }) async {
    final uri = Uri.parse('$baseUrl/api/graph/substitutions');
    
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'ingredient_id': ingredientId,
        if (context != null) 'context': context.toJson(),
        'limit': limit,
      }),
    );
    
    if (response.statusCode == 200) {
      return SubstitutionResponse.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to get substitutions: ${response.statusCode}');
    }
  }
  
  /// Get substitutions (simple GET method)
  Future<SubstitutionResponse> getSubstitutionsSimple({
    required String ingredientId,
    int limit = 10,
    String? form,
    String? dishType,
  }) async {
    final queryParams = {
      'limit': limit.toString(),
      if (form != null) 'form': form,
      if (dishType != null) 'dish_type': dishType,
    };
    
    final uri = Uri.parse('$baseUrl/api/graph/substitutions/$ingredientId')
        .replace(queryParameters: queryParams);
    
    final response = await http.get(uri);
    
    if (response.statusCode == 200) {
      return SubstitutionResponse.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to get substitutions: ${response.statusCode}');
    }
  }
  
  /// Disambiguate between commonly confused ingredients
  Future<ConfusionResult> resolveConfusion({
    required List<String> detectedIngredients,
    Map<String, dynamic>? visualFeatures,
    Map<String, dynamic>? userContext,
  }) async {
    final uri = Uri.parse('$baseUrl/api/graph/resolve-confusion');
    
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'detected_ingredients': detectedIngredients,
        if (visualFeatures != null) 'visual_features': visualFeatures,
        if (userContext != null) 'user_context': userContext,
      }),
    );
    
    if (response.statusCode == 200) {
      return ConfusionResult.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to resolve confusion: ${response.statusCode}');
    }
  }
  
  /// Get complementary ingredient pairing suggestions
  Future<PairingsResponse> getPairings({
    required List<String> ingredientIds,
    String? cuisineType,
    String? dishType,
    int limit = 20,
  }) async {
    final uri = Uri.parse('$baseUrl/api/graph/pairings');
    
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'ingredient_ids': ingredientIds,
        if (cuisineType != null) 'cuisine_type': cuisineType,
        if (dishType != null) 'dish_type': dishType,
        'limit': limit,
      }),
    );
    
    if (response.statusCode == 200) {
      return PairingsResponse.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to get pairings: ${response.statusCode}');
    }
  }
  
  /// Calculate recipe compatibility score
  Future<RecipeCompatibility> calculateRecipeCompatibility({
    required List<String> ingredientIds,
    String? cuisineType,
  }) async {
    final uri = Uri.parse('$baseUrl/api/graph/recipe-compatibility');
    
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'ingredient_ids': ingredientIds,
        if (cuisineType != null) 'cuisine_type': cuisineType,
      }),
    );
    
    if (response.statusCode == 200) {
      return RecipeCompatibility.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to calculate compatibility: ${response.statusCode}');
    }
  }
  
  /// Optimize grocery shopping list
  Future<GroceryOptimization> optimizeGroceryList({
    required List<String> ingredientIds,
    List<String>? userInventory,
    double? budgetConstraint,
  }) async {
    final uri = Uri.parse('$baseUrl/api/graph/optimize-grocery-list');
    
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'ingredient_ids': ingredientIds,
        if (userInventory != null) 'user_inventory': userInventory,
        if (budgetConstraint != null) 'budget_constraint': budgetConstraint,
      }),
    );
    
    if (response.statusCode == 200) {
      return GroceryOptimization.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to optimize grocery list: ${response.statusCode}');
    }
  }
  
  /// Record user feedback on substitution
  Future<void> recordSubstitutionFeedback({
    required String substitutionId,
    required bool wasAccepted,
    String? feedbackNote,
  }) async {
    final uri = Uri.parse('$baseUrl/api/graph/substitution-feedback');
    
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'substitution_id': substitutionId,
        'was_accepted': wasAccepted,
        if (feedbackNote != null) 'feedback_note': feedbackNote,
      }),
    );
    
    if (response.statusCode != 200) {
      throw Exception('Failed to record feedback: ${response.statusCode}');
    }
  }
  
  /// Health check
  Future<bool> checkHealth() async {
    try {
      final uri = Uri.parse('$baseUrl/api/graph/health');
      final response = await http.get(uri);
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }
}

// ============================================================================
// DATA MODELS
// ============================================================================

class SubstitutionContext {
  final String? form;
  final String? dishType;
  final String? cuisine;
  final List<String>? dietaryRestrictions;
  
  SubstitutionContext({
    this.form,
    this.dishType,
    this.cuisine,
    this.dietaryRestrictions,
  });
  
  Map<String, dynamic> toJson() {
    return {
      if (form != null) 'form': form,
      if (dishType != null) 'dish_type': dishType,
      if (cuisine != null) 'cuisine': cuisine,
      if (dietaryRestrictions != null) 'dietary_restrictions': dietaryRestrictions,
    };
  }
}

class SubstitutionResponse {
  final String ingredientId;
  final Map<String, dynamic>? context;
  final List<Substitution> substitutions;
  final int count;
  
  SubstitutionResponse({
    required this.ingredientId,
    this.context,
    required this.substitutions,
    required this.count,
  });
  
  factory SubstitutionResponse.fromJson(Map<String, dynamic> json) {
    return SubstitutionResponse(
      ingredientId: json['ingredient_id'],
      context: json['context'],
      substitutions: (json['substitutions'] as List)
          .map((s) => Substitution.fromJson(s))
          .toList(),
      count: json['count'],
    );
  }
}

class Substitution {
  final String substitutionId;
  final String ingredientId;
  final String canonicalName;
  final String category;
  final Map<String, dynamic> names;
  final List<String>? tasteProfile;
  final List<String>? commonUses;
  final String substitutionType;
  final double similarityScore;
  final List<String>? applicableForms;
  final List<String>? applicableDishes;
  final String? notes;
  final double? userAcceptanceRate;
  final Map<String, dynamic>? usageStats;
  
  Substitution({
    required this.substitutionId,
    required this.ingredientId,
    required this.canonicalName,
    required this.category,
    required this.names,
    this.tasteProfile,
    this.commonUses,
    required this.substitutionType,
    required this.similarityScore,
    this.applicableForms,
    this.applicableDishes,
    this.notes,
    this.userAcceptanceRate,
    this.usageStats,
  });
  
  factory Substitution.fromJson(Map<String, dynamic> json) {
    return Substitution(
      substitutionId: json['substitution_id'],
      ingredientId: json['ingredient_id'],
      canonicalName: json['canonical_name'],
      category: json['category'],
      names: Map<String, dynamic>.from(json['names']),
      tasteProfile: json['taste_profile'] != null
          ? List<String>.from(json['taste_profile'])
          : null,
      commonUses: json['common_uses'] != null
          ? List<String>.from(json['common_uses'])
          : null,
      substitutionType: json['substitution_type'],
      similarityScore: json['similarity_score'].toDouble(),
      applicableForms: json['applicable_forms'] != null
          ? List<String>.from(json['applicable_forms'])
          : null,
      applicableDishes: json['applicable_dishes'] != null
          ? List<String>.from(json['applicable_dishes'])
          : null,
      notes: json['notes'],
      userAcceptanceRate: json['user_acceptance_rate']?.toDouble(),
      usageStats: json['usage_stats'],
    );
  }
  
  /// Get quality indicator for substitution
  String getQualityLevel() {
    if (similarityScore >= 0.9) return 'Excellent';
    if (similarityScore >= 0.75) return 'Good';
    if (similarityScore >= 0.6) return 'Fair';
    return 'Emergency Only';
  }
}

class ConfusionResult {
  final bool needsDisambiguation;
  final int detectedCount;
  final int? confusionPatterns;
  final List<ConfusionRecommendation>? recommendations;
  final String? message;
  
  ConfusionResult({
    required this.needsDisambiguation,
    required this.detectedCount,
    this.confusionPatterns,
    this.recommendations,
    this.message,
  });
  
  factory ConfusionResult.fromJson(Map<String, dynamic> json) {
    return ConfusionResult(
      needsDisambiguation: json['needs_disambiguation'],
      detectedCount: json['detected_count'],
      confusionPatterns: json['confusion_patterns'],
      recommendations: json['recommendations'] != null
          ? (json['recommendations'] as List)
              .map((r) => ConfusionRecommendation.fromJson(r))
              .toList()
          : null,
      message: json['message'],
    );
  }
}

class ConfusionRecommendation {
  final List<String> confusedPair;
  final String reason;
  final List<String> disambiguationTips;
  final List<String> visualDifferences;
  final String confidence;
  final String? matchedVisualClue;
  
  ConfusionRecommendation({
    required this.confusedPair,
    required this.reason,
    required this.disambiguationTips,
    required this.visualDifferences,
    required this.confidence,
    this.matchedVisualClue,
  });
  
  factory ConfusionRecommendation.fromJson(Map<String, dynamic> json) {
    return ConfusionRecommendation(
      confusedPair: List<String>.from(json['confused_pair']),
      reason: json['reason'],
      disambiguationTips: List<String>.from(json['disambiguation_tips']),
      visualDifferences: List<String>.from(json['visual_differences']),
      confidence: json['confidence'],
      matchedVisualClue: json['matched_visual_clue'],
    );
  }
}

class PairingsResponse {
  final List<String> inputIngredients;
  final String? cuisineType;
  final String? dishType;
  final List<IngredientPairing> pairings;
  final int count;
  
  PairingsResponse({
    required this.inputIngredients,
    this.cuisineType,
    this.dishType,
    required this.pairings,
    required this.count,
  });
  
  factory PairingsResponse.fromJson(Map<String, dynamic> json) {
    return PairingsResponse(
      inputIngredients: List<String>.from(json['input_ingredients']),
      cuisineType: json['cuisine_type'],
      dishType: json['dish_type'],
      pairings: (json['pairings'] as List)
          .map((p) => IngredientPairing.fromJson(p))
          .toList(),
      count: json['count'],
    );
  }
}

class IngredientPairing {
  final String pairingId;
  final String ingredientId;
  final String canonicalName;
  final String category;
  final Map<String, dynamic> names;
  final List<String>? tasteProfile;
  final double pairingScore;
  final String pairingType;
  final List<String> cuisineTypes;
  final List<String> dishTypes;
  final String source;
  final int timesUsedTogether;
  
  IngredientPairing({
    required this.pairingId,
    required this.ingredientId,
    required this.canonicalName,
    required this.category,
    required this.names,
    this.tasteProfile,
    required this.pairingScore,
    required this.pairingType,
    required this.cuisineTypes,
    required this.dishTypes,
    required this.source,
    required this.timesUsedTogether,
  });
  
  factory IngredientPairing.fromJson(Map<String, dynamic> json) {
    return IngredientPairing(
      pairingId: json['pairing_id'],
      ingredientId: json['ingredient_id'],
      canonicalName: json['canonical_name'],
      category: json['category'],
      names: Map<String, dynamic>.from(json['names']),
      tasteProfile: json['taste_profile'] != null
          ? List<String>.from(json['taste_profile'])
          : null,
      pairingScore: json['pairing_score'].toDouble(),
      pairingType: json['pairing_type'],
      cuisineTypes: List<String>.from(json['cuisine_types']),
      dishTypes: List<String>.from(json['dish_types']),
      source: json['source'],
      timesUsedTogether: json['times_used_together'],
    );
  }
  
  /// Get pairing strength indicator
  String getPairingStrength() {
    if (pairingScore >= 0.9) return 'Perfect Match';
    if (pairingScore >= 0.8) return 'Excellent';
    if (pairingScore >= 0.7) return 'Very Good';
    if (pairingScore >= 0.6) return 'Good';
    return 'Fair';
  }
}

class RecipeCompatibility {
  final double compatibilityScore;
  final String compatibilityLevel;
  final String confidence;
  final int ingredientCount;
  final int knownPairings;
  final List<PairingDetail> pairingDetails;
  
  RecipeCompatibility({
    required this.compatibilityScore,
    required this.compatibilityLevel,
    required this.confidence,
    required this.ingredientCount,
    required this.knownPairings,
    required this.pairingDetails,
  });
  
  factory RecipeCompatibility.fromJson(Map<String, dynamic> json) {
    return RecipeCompatibility(
      compatibilityScore: json['compatibility_score'].toDouble(),
      compatibilityLevel: json['compatibility_level'],
      confidence: json['confidence'],
      ingredientCount: json['ingredient_count'],
      knownPairings: json['known_pairings'],
      pairingDetails: (json['pairing_details'] as List)
          .map((p) => PairingDetail.fromJson(p))
          .toList(),
    );
  }
}

class PairingDetail {
  final List<String> ingredients;
  final double score;
  final String type;
  final List<String>? cuisines;
  
  PairingDetail({
    required this.ingredients,
    required this.score,
    required this.type,
    this.cuisines,
  });
  
  factory PairingDetail.fromJson(Map<String, dynamic> json) {
    return PairingDetail(
      ingredients: List<String>.from(json['ingredients']),
      score: json['score'].toDouble(),
      type: json['type'],
      cuisines: json['cuisines'] != null
          ? List<String>.from(json['cuisines'])
          : null,
    );
  }
}

class GroceryOptimization {
  final int totalIngredients;
  final int alreadyHave;
  final int needToBuy;
  final List<dynamic> itemsToBuy;
  final Map<String, dynamic> optimizations;
  final int? estimatedSavings;
  
  GroceryOptimization({
    required this.totalIngredients,
    required this.alreadyHave,
    required this.needToBuy,
    required this.itemsToBuy,
    required this.optimizations,
    this.estimatedSavings,
  });
  
  factory GroceryOptimization.fromJson(Map<String, dynamic> json) {
    return GroceryOptimization(
      totalIngredients: json['total_ingredients'],
      alreadyHave: json['already_have'],
      needToBuy: json['need_to_buy'],
      itemsToBuy: json['items_to_buy'],
      optimizations: json['optimizations'],
      estimatedSavings: json['estimated_savings'],
    );
  }
}
